cmd/snap/cmd_buy.go

```diff
@@ -117,33 +110,28 @@ Once completed, return here and run 'snap buy %s' again.`), snap.Name)
 		return err
 	}
 
-	reader := bufio.NewReader(nil)
-	reader.Reset(Stdin)
+	fmt.Fprintf(Stdout, i18n.G(`Please re-enter your Ubuntu One password to purchase '%s' from '%s' for %s. Press ctrl-c to cancel.`),
+		snap.Name, snap.Developer, formatPrice(opts.Price, opts.Currency))
+	fmt.Fprint(Stdout, "\n")
 
-	fmt.Fprintf(Stdout, i18n.G("Do you want to buy %q from %q for %s? (Y/n): "), snap.Name,
-		snap.Developer, formatPrice(opts.Price, opts.Currency))
-
-	response, _, err := reader.ReadLine()
+	user, err := client.ReadAuthData()
```

Comment:
The re-authentication you want to do is against the store. What we have in auth.json is the authentication with snapd, though, and we don't want to throw that out and redo it without context. The email you're picking up from it is the email that snapd itself returned, and it still knows it internally. So there's no reason for this code path to be reading auth.json out-of-band.